src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -39,107 +37,84 @@ import kotlin.coroutines.Continuation
  */
 internal fun enumerateObjects(obj: Any): Map<Any, Int> {
     val objectNumberMap = hashMapOf<Any, Int>()
-    enumerateObjects(obj, Collections.newSetFromMap(IdentityHashMap()), objectNumberMap)
-
+    enumerateObjects(obj, objectNumberMap)
     return objectNumberMap
 }
 
 /**
  * Recursively traverses an object to enumerate it and all nested objects.
  *
  * @param obj object to traverse
- * @param processedObjects a set of already processed objects. Required in case of cyclic references.
  * @param objectNumberMap result enumeration map
  */
-private fun enumerateObjects(obj: Any, processedObjects: MutableSet<Any>, objectNumberMap: MutableMap<Any, Int>) {
+private fun enumerateObjects(obj: Any, objectNumberMap: MutableMap<Any, Int>) {
     if (obj is Class<*> || obj is ClassLoader) return
-    if (!processedObjects.add(obj)) return
-    val objectNumber = getObjectNumber(obj.javaClass, obj)
-    objectNumberMap[obj] = objectNumber
-
-    var clazz: Class<*>? = obj.javaClass
-    if (clazz!!.isArray) {
-        if (obj is Array<*>) {
-            obj.forEach { element -> element?.let { enumerateObjects(it, processedObjects, objectNumberMap) } }
+    objectNumberMap[obj] = getObjectNumber(obj.javaClass, obj)
+
+    val processObject: (Any?) -> Any? = processObject@{ value: Any? ->
+        if (value == null || value is Class<*> || value is ClassLoader) return@processObject null
+
+        // we jump through most of the atomic classes
+        var jumpObj: Any? = value
+
+        if (isAtomic(jumpObj)) {
```

Comment:
TODO: replace with `isAtomicClass(...)` using string comparison with classname. Due to a possibility to extend from `AtomicReference<*>` for example. So if we have `Node : AtomicReference<*>()` we want to traverse node's fields

Reply:
TODO: also handle nested atomic references case:  `AtomicReference<AtomicReference<...>>`